File / FileCache

bytinbit opened the pull request:
RSVP for the structure of the File-Class, it determines everything else.

Commit:
skeleton, draft of file-class

Commit:
skeleton refined, comments added, basic structure

Commit:
changed intial constructor structure to dictionary

Commit:
design decision: all File properties are stored in a dict, since it facilitates the conversion to json later for the FileCache.

Commit:
write/retrieve file from filecache draft

The-Compiler commented on src/kitovu/sync/filecache.py:
```diff
@@ -0,0 +1,70 @@
+        # write File's cached_digest and path
+        pass
+
+    def discover_changes_in_file(self, file: File) -> int:
```
Various methods (e.g. this one) take a `File` object. However, to construct a `File` you need a lot of information only the `FileCache` knows but the caller doesn't (such as `local_digest_at_synctime`), so that's not going to work. I think it'd be best to pass a local file path there (i.e. the path where the file is actually stored on disk), but I don't have a clear picture of things yet, either. I'd say let's do that for now.

The-Compiler commented on src/kitovu/sync/filecache.py:
```diff
@@ -0,0 +1,70 @@
+
+
+class FileCache:
+    FILECACHE: str = "filecache.json"
```
It'd be good if the filename (and path) could be passed to the constructor (`__init__`) instead. That'll make it easier to test as you can use a `FileCache` with a temporary file easily.

The-Compiler commented on src/kitovu/sync/filecache.py:
```diff
@@ -0,0 +1,70 @@
+    def __init__(self, local_digest_at_synctime: str,
+                 relative_path_with_filename: pathlib.PurePath,
+                 plugin: syncplugin.AbstractSyncPlugin) -> None:
+        file_properties = {"path": relative_path_with_filename,
```
I don't understand what you're trying to do here. All this data should probably be saved in separate attributes (`self._relative_path_with_filename`, `self._local_digest_at_synctime`, etc.).

If you need some dict-structure like this for JSON, you can add a `to_dict()` method, but it doesn't make much sense to store it like that internally I think.

bytinbit replied:
Initially I had this construct as you suggested, then I changed it, since I had the idea, that it is easy to parse a dict to json. I'll revert to the original idea then.

The-Compiler commented on src/kitovu/sync/filecache.py:
```diff
@@ -0,0 +1,70 @@
+class FileCache:
+    FILECACHE: str = "filecache.json"
+
+    def write_file_to_filecache(self, file: File):
```
You seem to try reading/writing the JSON with every operation - that's more complexity than needed, and also won't work out (you can't just dump a file to JSON and append that to the file).

Instead, I imagine things to work something like this:

- Read the entire file cache from disk once (with `json.load`) with a method like `FileCache.read()` (shouldn't need any arguments)
- Interact with the `FileCache` by adding files to it, or comparing them, etc.
- Write it to disk (with `json.dump`) once everything is finished with a method like `FileCache.write()` (or `.store()`) - again, probably without any arguments.

The-Compiler commented on src/kitovu/sync/filecache.py:
```diff
@@ -0,0 +1,70 @@
+            with self.FILECACHE.open("a+") as f:  # a+ = if it exsists, append; if not, create a new one
+                f.write(file_to_json)
+        except OSError as error:
+            raise utils.UsageError(f"Could not write the file, {error}")
```
I wouldn't add error handling for this right now - it's very unlikely to happen (the file cache should always be writable), and a proper exception makes things much easier to debug than a "Could not write the file" message out of context (what file?). We can still discuss whether this kind of thing is needed at all, but for a first iteration, don't worry about it.

Commit:
comments integrated: File-constructor not a dict, but individual entries; load_filecache returns deserialised list of files; write_filecache_to_json writes a list of files to disk.

The-Compiler commented on src/kitovu/sync/kitovu_filecache.py:
```diff
@@ -0,0 +1,68 @@
+        self._relative_path_with_filename: pathlib.PurePath
+        self._plugin = plugin
+        self._remote_digest = ""
+        self._local_digest = ""
```
You'll probably still need a `to_dict()` method or so, which converts this to a dictionary like the one you had before. My main point is that it's cleaner to store the internal data like this.

The-Compiler commented on src/kitovu/sync/kitovu_filecache.py:
```diff
@@ -0,0 +1,68 @@
+
+
+class FileCache:
+    def __init__(self, filename: str, path_to_json: pathlib.Path):
```
What's the difference between `filename` and `path_to_json`? Would it suffice to only have one path (which is the full path to the file)?

The-Compiler commented on src/kitovu/sync/kitovu_filecache.py:
```diff
@@ -0,0 +1,68 @@
+        self._FILECACHE = filename  # filecache.json
+        self._path = path_to_json
+
+    def write_filecache_to_jsonfile(self, file: typing.List[File]):
```
This might work, but is not what I meant exactly earlier - basically, I'd say the `FileCache` manages itself (and the caller can add individual files), and then you can tell it "save yourself" and "restore yourself".

Commit:
Merge remote-tracking branch 'origin/master' into feature-filecache

# Conflicts:
#	src/kitovu/sync/syncing.py

Commit:
restructured FileCache, File-Class; write()-FileCache added

Commit:
FileCache-load() completed

Commit:
first draft of conflict handling when remote/local/cached differs

Commit:
first draft to integrate FileCache into sync-process

Commit:
first version: sync-process adapted to FileCache; update() to update FileCache

Commit:
first draft to check,

The-Compiler commented on src/kitovu/sync/kitovu_filecache.py:
```diff
@@ -0,0 +1,113 @@
+        self._filename: pathlib.Path = filename
+        self._data: typing.Dict[pathlib.Path, File] = {}
+
+    def _compare_digests(self, remote_digest: str, local_digest: str, cached_digest: str) -> Filestate:
```
You could make this a bit nicer to read with two helper variables:

```python
local_changed: bool = local_digest != cached_digest
remote_changed: bool = remote_digest != cached_digest
```

The-Compiler commented on src/kitovu/sync/kitovu_filecache.py:
```diff
@@ -0,0 +1,113 @@
+            digest: str = value["digest"]
+            plugin: str = value["plugin"]
+            self._data[pathlib.Path(key)] = File(local_digest_at_synctime=digest, plugin=plugin)
+            # FIXME should be of type AbstractSyncPlugin - conversion needed?
```
You'll need to get the plugin instance from the name here, i.e. what happens in `syncing._find_plugin`. However, that'd introduce a circular dependency (sycing depends on filecache depends on syncing), so `_find_plugin` should be moved to another place (like the `kitovu.sync.syncplugin` module).

The-Compiler commented on src/kitovu/sync/kitovu_filecache.py:
```diff
@@ -0,0 +1,113 @@
+            self._data[pathlib.Path(key)] = File(local_digest_at_synctime=digest, plugin=plugin)
+            # FIXME should be of type AbstractSyncPlugin - conversion needed?
+
+    def update(self, path: pathlib.Path, plugin: syncplugin.AbstractSyncPlugin, local_digest_at_synctime):
```
"update" usually means "change an entry which is already there" - wouldn't `insert` be a better name?

The-Compiler replied:
`local_digest_at_synctime` has no type.

bytinbit replied:
the method is used when A has become A', and we need to reflect this in the FileCache - hence ``update()`` or ``modify()``, the latter which you originally suggested, are more concise. I'll switch to ``modify()``, since you suggested this earlier on.

The-Compiler commented on src/kitovu/sync/kitovu_filecache.py:
```diff
@@ -0,0 +1,113 @@
+
+    def update(self, path: pathlib.Path, plugin: syncplugin.AbstractSyncPlugin, local_digest_at_synctime):
+        file = File(local_digest_at_synctime=local_digest_at_synctime, plugin=plugin)
+        self._data.update(path, file)
```
Calling `.update()` on a dict adds another dict to it:

```python
>>> d1 = {'one': '一', 'two': '二'}
>>> d2 = {'three': '三', 'four': '四'}
>>> d1.update(d2)
>>> d1
{'one': '一', 'two': '二', 'three': '三', 'four': '四'}
```

I think you simply want `self._data[path] = file` here?

The-Compiler commented on src/kitovu/sync/kitovu_filecache.py:
```diff
@@ -0,0 +1,113 @@
+
+    def discover_changes(self, path: pathlib.Path, plugin: syncplugin.AbstractSyncPlugin) -> Filestate:
+        """checks if the file that is currently downloaded (path-argument) has changed
+        in comparison to the local filecache and local file)."""
```
The typical format for docstrings is similar to a git commit (or email with subject/body) - a one-line summary, blank line, and then a longer explanation, all in imperative form:

```
"""Check if the file currently being downloaded has changed.

(Longer text here)
"""
```

Also see [PEP 257](https://www.python.org/dev/peps/pep-0257/) which is basically PEP8 but for docstrings.

The-Compiler commented on src/kitovu/sync/kitovu_filecache.py:
```diff
@@ -0,0 +1,113 @@
+
+        # use path as index into already loaded _data
+        file: File = self._data[path]
+        cached_digest = file["digest"]
```
type?

The-Compiler commented on src/kitovu/sync/kitovu_filecache.py:
```diff
@@ -0,0 +1,113 @@
+
+        # get cached_digest, compare it to file that is synchronised
+        remote_digest = plugin.create_remote_digest(path)
+        local_digest = plugin.create_local_digest(path)
```
types? :wink:

The-Compiler commented on src/kitovu/sync/kitovu_filecache.py:
```diff
@@ -0,0 +1,113 @@
+        file = File(local_digest_at_synctime=local_digest_at_synctime, plugin=plugin)
+        self._data.update(path, file)
+
+    def discover_changes(self, path: pathlib.Path, plugin: syncplugin.AbstractSyncPlugin) -> Filestate:
```
Can't we get the plugin from the `File` object?

The-Compiler commented on src/kitovu/sync/syncing.py:
```diff
@@ -10,6 +10,7 @@
 from kitovu.sync import syncplugin
 from kitovu.sync.settings import Settings, ConnectionSettings
 from kitovu.sync.plugin import smb
+from kitovu.sync import kitovu_filecache
```
Why is the file named `kitovu_filecache` by the way? For consistency, you should drop that prefix. Make sure to rename it in a git commit without any other changes though, that makes things easier to follow.

The-Compiler commented on src/kitovu/sync/syncing.py:
```diff
@@ -29,19 +30,23 @@ def _find_plugin(pluginname: str) -> syncplugin.AbstractSyncPlugin:
     plugin: syncplugin.AbstractSyncPlugin = manager.driver
     return plugin
 
-
```
Unrelated whitespace change (two blank lines between top-level functions which aren't in a class)

The-Compiler commented on src/kitovu/sync/syncing.py:
```diff
@@ -29,19 +30,23 @@ def _find_plugin(pluginname: str) -> syncplugin.AbstractSyncPlugin:
     plugin.configure(connection_settings.connection)
     plugin.connect()
+    filecache: kitovu_filecache.FileCache = kitovu_filecache.FileCache(pathlib.PurePath("filecache.json"))
+    # FIXME add path from settings instead of filecache.json
```
You should be able to pass something like `pathlib.Path(appdirs.user_data_dir('kitovu')) / 'filecache.json'`.

(`Path` and not `PurePath` because it's a file on disk)

The-Compiler commented on src/kitovu/sync/syncing.py:
```diff
@@ -51,17 +56,51 @@ def start(connection_settings: ConnectionSettings) -> None:
+    # 2. remote deleted (triggers exception), local exists AND changed (local_digest and cached_digest differ) => BOTH*
+
+            # if file doesn't exist: case 3 (normal case)
+            if not pathlib.Path(local_path / item.relative_to(remote_path)).exists():
```
There's a lot of duplication with code handling `Filestate.REMOTE` here, which could be simplified. Couldn't the cache check whether the local file exists and return `Filestate.REMOTE` if it doesn't?

Commit:
code review for filecache: first batch of commentaries resolved

Commit:
kitovu_filecache.py renamed to filecache.py

The-Compiler commented on src/kitovu/sync/filecache.py:
```diff
@@ -0,0 +1,115 @@
+
+    def _compare_digests(self, remote_digest: str, local_digest: str, cached_digest: str) -> Filestate:
+        local_changed: bool = local_digest != cached_digest
+        remote_changed: bool = remote_digest != cached_digest
```
You could rewrite the `==` parts as well, using e.g. `elif remote_changed and not local_changed:` (case 4) or `if not remote_changed and not local_changed:` (case 5), etc.

Commit:
reflects weekly meeting's decisions that need to be further implemented

Commit:
compare_digests refactored

Commit:
property "name" added to AbstractSyncPlugin-Cass

Commit:
nice docstring for file cache

Commit:
changed AbstractSyncPlugin-property from abc-name to string-name

Commit:
file-property "plugin" changed from abstract to string-name

Commit:
load_plugin() eliminated

Commit:
refactored syncing.start()

Commit:
FILESTATE names changed, code in syncing.py refactored

Commit:
code refactored, comments added for clarification, included remote_full_path to discover changes (previously: wrong path used)

Commit:
comments cleaned up, code refactored in filecache

Commit:
further refactoring, two superfluous variables in File-Class deleted

Commit:
setup of filecache tests

Commit:
dummyplugin has now name, first test for File-Class

Commit:
file-class attribute "relative_path_with_filename" purged, not needed anymore since we handle the identification of its path via FileCache

Commit:
Test Load/Write of FileCache completed

Commit:
further tests TestChange completed

The-Compiler commented on src/kitovu/sync/filecache.py:
```diff
@@ -0,0 +1,124 @@
+
+    def load(self) -> None:
+        """This is called first when the synchronisation process is started."""
+        with self._filename.open("r") as f:
```
This currently crashes when the filecache doesn't exist initially.

bytinbit replied:
fixed in syncing.py

The-Compiler commented on src/kitovu/sync/filecache.py:
```diff
@@ -0,0 +1,124 @@
+        if not local_full_path.exists():
+            return FileState.NEW
+
+        file: File = self._data[local_full_path]
```
We should probably handle the case where the local file exists (because the user downloaded it manually or whatever) but it's not in the cache. Not sure how to handle that, though.

bytinbit replied:
I've added this as issue EPJ-79 to our backlog.

The-Compiler commented on src/kitovu/sync/filecache.py:
```diff
@@ -0,0 +1,124 @@
+            return FileState.NEW
+
+        file: File = self._data[local_full_path]
+        cached_digest: str = file["digest"]
```
This should be `file.cached_digest` probably? (as it's a `File` object, not a dict)

bytinbit replied:
true, fixed.

Commit:
fixed: kitovu crashed when the file didn't exist initially.

Commit:
fixed missing class-variable when using cached_digest

Commit:
further tests, dummyplugin somewhat adapted

Commit:
resolved: filecache.json doesn't crash when it doesn't exist initially

Commit:
write() of filecache-exception fixed when filecache.json doesn't exist initially

Commit:
dummy-plugin fixutre purged, moved to conftest.py

Commit:
Code Refactored to use temppath for the DummyPlugin, not hard-coded paths

Commit:
all tests completed and green

Commit:
final test for wrong plugin name

The-Compiler committed:
Fix helpers import

The-Compiler committed:
Fix whitespace issues

The-Compiler committed:
Rewrite TestFileState to use parametrization

The-Compiler committed:
Rewrite FileCache comment

The-Compiler committed:
Clean up import

The-Compiler committed:
Merge branch 'master' into feature-filecache

The-Compiler commented:
Oh, looks like there's some issues with tests on Windows, and pylint/mypy. Will fix them tomorrow, but please take a look whenever you have the time to get some free code feedback from those tools :wink:

The-Compiler committed:
Break long lines

The-Compiler committed:
Use named arguments for cache.discover_changes call

That way we can be sure we don't mix things up

The-Compiler committed:
Turn off missing-docstring for pylint

Already done on master with the GUI branch merged

The-Compiler committed:
Add an "else:" branch to _compare_digests

This should never happen because we do handle all possible cases above, but
if we don't do that, all the tools which don't know yell at us.

The-Compiler committed:
pylint: Allow to use "f" as variable name

It's commonly used for file objects

The-Compiler committed:
Fix import order

attr and appdirs are third-party libraries

The-Compiler committed:
Add missing return type annotations

The-Compiler committed:
Make sure plugins have a name

This also makes mypy happy, which is more clever than I thought!

The-Compiler committed:
Also create parent directories when creating filecache dir

On Windows, we get a path like this from appveyor:
C:/Users/appveyor/AppData/Local/kitovu/kitovu/filecache.json

There, only AppData/Local/ exists, but the other parts don't, so we need to add
parents=True to create them too.

The-Compiler commented:
Fixed everything up and merged! :tada: Please take a look at the commits I pushed above when you get the time, but it's mostly small stuff.
